sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/BodyMemory.cs

```diff
@@ -6,96 +6,144 @@
     {
-        private ArrayBufferWriter<byte> _writer;
-        private IList<ReadOnlyMemory<byte>> _segments;
-        private IEnumerable<ReadOnlyMemory<byte>> _lazySegments;
-
-        private ReadOnlyMemory<byte> WrittenMemory
+        public static BodyMemory FromReadOnlyMemory(IEnumerable<ReadOnlyMemory<byte>> segments)
         {
-            get
+            return segments switch
             {
-                if (_lazySegments != null)
-                {
-                    foreach (var segment in _lazySegments)
-                    {
-                        Append(segment);
-                    }
-
-                    _lazySegments = null;
-                }
-
-                return _writer?.WrittenMemory ?? ReadOnlyMemory<byte>.Empty;
-            }
+                BodyMemory bodyMemory => bodyMemory,
+                _ => new SendWithMultipleDataSegmentsBodyMemory(segments)
+            };
         }
 
-        // for the send path the combined memory for the body is not precomputed
-        private BodyMemory(IEnumerable<ReadOnlyMemory<byte>> dataSegments)
+        public static BodyMemory FromReadOnlyMemory(ReadOnlyMemory<byte> segment)
         {
-            _lazySegments = dataSegments;
+            return new SendWithSingleDataSegmentBodyMemory(segment);
         }
 
-        // for the receive path the combined memory is precomputed because we need to copy the underlying AMQP data anyway
-        private BodyMemory(IEnumerable<Data> dataSegments)
+        public static BodyMemory FromAmqpData(IEnumerable<Data> segments)
         {
-            foreach (var segment in dataSegments)
-            {
-                Append(segment);
-            }
+            return new ReceiveBodyMemory(segments ?? Enumerable.Empty<Data>());
         }
 
-        public static BodyMemory FromReadOnlyMemory(IEnumerable<ReadOnlyMemory<byte>> segments)
-        {
-            return segments as BodyMemory ?? new BodyMemory(segments);
-        }
+        protected abstract ReadOnlyMemory<byte> WrittenMemory { get; }
 
-        public static BodyMemory FromAmqpData(IEnumerable<Data> segments)
+        public abstract IEnumerator<ReadOnlyMemory<byte>> GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator()
         {
-            return new BodyMemory(segments ?? Enumerable.Empty<Data>());
+            return GetEnumerator();
         }
 
-        public IEnumerator<ReadOnlyMemory<byte>> GetEnumerator()
+        public static implicit operator ReadOnlyMemory<byte>(BodyMemory memory)
         {
-            return _segments?.GetEnumerator() ?? _lazySegments.GetEnumerator();
+            return memory.WrittenMemory;
         }
 
-        IEnumerator IEnumerable.GetEnumerator()
+        private sealed class SendWithSingleDataSegmentBodyMemory : BodyMemory
         {
-            return GetEnumerator();
-        }
+            public SendWithSingleDataSegmentBodyMemory(ReadOnlyMemory<byte> dataSegment)
```

Comment:
nit: can we keep both send classes together, i.e. one after the other (right now the receive class is in between).

Reply:
Now that I think about it, the SendWithMultiple class actually can apply for both sent and received messages since the body can still be mutated in received messages. Maybe we can name it something like AmqpMessageModifiedBodyMemory?

Reply:
I see what I can do today and think about better names too

Reply:
have a look. I pushed a few updates. By the way if you don't like the test names or want to do small tweaks I gave edit rights on this PR.